lib/Conversion/TorchToTosa/TosaLegalizeCommon.cpp

```diff
@@ -27,6 +28,370 @@
+
+  auto paramsType = paramsValue.getType().dyn_cast<RankedTensorType>();
+  auto indexType = indexValue.getType().dyn_cast<RankedTensorType>();
+  auto paramsShape = paramsType.getShape(); // [1 4 3]
+  auto indexShape = indexType.getShape();   // [1 4 2]
+  int paramsRank = paramsShape.size();      // 3
+  int indexRank = indexShape.size();        // 3
+
+  // Initialize the final tf indices shape, and the shape of each dim that can
+  // concat to this tf indices
+  SmallVector<int64_t> indicesShape;       // [1 4 2 3]
+  SmallVector<int64_t> indicesOneDimShape; // [1 4 2 1]
+  for (auto shape : indexShape) {
+    indicesShape.push_back(shape);
+    indicesOneDimShape.push_back(shape);
+  }
+  indicesShape.push_back(paramsRank);
+  indicesOneDimShape.push_back(1);
+
+  // Get the chosen axis index
+  // indexValue reshape to indicesDim: shape append 1
+  // [1 4 2] -> [1 4 2 1]
+  // dim2:	tensor([[   [ [0], [0] ],
+  //			    [ [1], [0] ],
+  //			    [ [2], [1] ],
+  //			    [ [2], [1] ], ]]) 1*4*2*1
+  auto indicesChosenAxis = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+      rewriter, op->getLoc(),
+      GetTypeFromTensorShape(indicesOneDimShape, indexType.getElementType()),
+      indexValue, rewriter.getI64ArrayAttr(indicesOneDimShape));
+
+  SmallVector<Value> concatInputs;
+  for (auto dim = 0; dim < paramsRank; dim++) {
+    if (dim != axis) {
+      auto indices = createOneDimTfIndices(rewriter, op,
+                                           /*result shape*/ indicesOneDimShape,
+                                           dim, indexShape);
+      concatInputs.push_back(indices.value());
+    } else {
+      // the chosen axis indices will be replaced by index[i][j][k]
+      concatInputs.push_back(indicesChosenAxis.getResult());
+    }
+  }
+
+  // detailed example explanation
+  // https://gist.github.com/AmosLewis/932a8dee3ba7657dcc6d09a4da4775d4 Get TF
+  // indices: 1*4*2*3
+  // [[  d0 d1  d2  d0 d1 d2
+  //    [[0, 0, 0],[0, 0, 0]],
+  //    [[0, 1, 1],[0, 1, 0]],
+  //    [[0, 2, 2],[0, 2, 1]],
+  //    [[0, 3, 2],[0, 3, 1]]
+  // ]]
+  auto indicesTf = tosa::CreateOpAndInfer<tosa::ConcatOp>(
+      rewriter, op->getLoc(),
+      GetTypeFromTensorShape(indicesShape, rewriter.getIntegerType(32)),
+      concatInputs, indexRank);
+
+  return indicesTf.getResult();
+}
+
+// Lowers Gather operators to a sequence of TOSA ops.
+// taken from
+// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc
+llvm::Optional<Value> convertGatherNdOp(PatternRewriter &rewriter,
+                                        Operation *op, Type out_type,
+                                        Value params_value,
+                                        Value indices_value) {
+  auto result_type = out_type.dyn_cast<ShapedType>();
+  auto params_type = params_value.getType().dyn_cast<RankedTensorType>();
+  auto indices_type = indices_value.getType().dyn_cast<RankedTensorType>();
```

Comment:
In torch-mlir, variables names are always camel cased (note: I'm just highlighting these, but there's more cases in this PR)

Reply:
Done